README.md
# Passport-Remember Me

[Passport](http://passportjs.org/) strategy for authenticating based on a
remember me cookie.

This module lets you authenticate using a remember me cookie (aka persistent
login) in your Node.js applications.  By plugging into Passport, remember me
authentication can be easily and unobtrusively integrated into any application
or framework that supports [Connect](http://www.senchalabs.org/connect/)-style
middleware, including [Express](http://expressjs.com/).

## Install

    $ npm install passport-remember-me

## Usage

#### Configure Strategy

The remember me authentication strategy authenticates users using a token stored
in a remember me cookie.  The strategy requires a `verify` callback, which
consumes the token and calls `done` providing a user.

The strategy also requires an `issue` callback, which issues a new token.  For
security reasons, remember me tokens should be invalidated after being used.
The `issue` callback supplies a new token that will be stored in the cookie for
next use.

    passport.use(new RememberMeStrategy(
      function(token, done) {
        Token.consume(token, function (err, user) {
          if (err) { return done(err); }
          if (!user) { return done(null, false); }
          return done(null, user);
        });
      },
      function(user, done) {
        var token = utils.generateToken(64);
        Token.save(token, { userId: user.id }, function(err) {
          if (err) { return done(err); }
          return done(null, token);
        });
      }
    ));

#### Authenticate Requests

Use `passport.authenticate()`, specifying the `'remember-me'` strategy, to
authenticate requests.

This is typically used in an application's middleware stack, to log the user
back in the next time they visit any page on your site.  For example:

    app.configure(function() {
      app.use(express.cookieParser());
      app.use(express.bodyParser());
      app.use(express.session({ secret: 'keyboard cat' }));
      app.use(passport.initialize());
      app.use(passport.session());
      app.use(passport.authenticate('remember-me'));
      app.use(app.router);
    });
    
Note that `passport.session()` should be mounted *above* `remember-me`
authentication, so that tokens aren't exchanged for currently active login
sessions.

#### Setting the Remember Me Cookie

If the user enables "remember me" mode, an initial cookie should be set when
they login.

    app.post('/login', 
      passport.authenticate('local', { failureRedirect: '/login', failureFlash: true }),
      function(req, res, next) {
        // issue a remember me cookie if the option was checked
        if (!req.body.remember_me) { return next(); }
    
        var token = utils.generateToken(64);
        Token.save(token, { userId: req.user.id }, function(err) {
          if (err) { return done(err); }
          res.cookie('remember_me', token, { path: '/', httpOnly: true, maxAge: 604800000 }); // 7 days
          return next();
        });
      },
      function(req, res) {
        res.redirect('/');
      });

#### Security Considerations

If not managed correctly, using a "remember me" cookie for automatic
authentication increases a service's exposure to potential security threats.
There are a number of techniques to reduce and mitigate these threats, and it
is a matter of application-level policy to asses the level of risk and implement
appropriate counter measures.

The following list is recommended reading for understanding these risks:

- [The definitive guide to forms based website authentication](http://stackoverflow.com/questions/549/the-definitive-guide-to-forms-based-website-authentication)
- [Persistent Login Cookie Best Practice](http://fishbowl.pastiche.org/2004/01/19/persistent_login_cookie_best_practice/)
- [Improved Persistent Login Cookie Best Practice](http://jaspan.com/improved_persistent_login_cookie_best_practice) [(archive)](http://web.archive.org/web/20130214051957/http://jaspan.com/improved_persistent_login_cookie_best_practice)

## Tests

    $ npm install
    $ make test

[![Build Status](https://secure.travis-ci.org/jaredhanson/passport-remember-me.png)](http://travis-ci.org/jaredhanson/passport-remember-me)

## Credits

  - [Jared Hanson](http://github.com/jaredhanson)

## License

[The MIT License](http://opensource.org/licenses/MIT)

Copyright (c) 2013 Jared Hanson <[http://jaredhanson.net/](http://jaredhanson.net/)>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.3](https://github.com/gospime/passport-remember-me/compare/v0.1.2...v0.1.3) (2019-12-01)


### Bug Fixes

* minor ([a3190db](https://github.com/gospime/passport-remember-me/commit/a3190dbdda31659610e4e78e6593fa1b60421fc7))

### [0.1.2](https://github.com/gospime/passport-remember-me/compare/v0.1.1...v0.1.2) (2019-12-01)


### Bug Fixes

* corrected critical bug - undefined value ([22a6b51](https://github.com/gospime/passport-remember-me/commit/22a6b518b4a4452868d5d5f7bfde9f9b10c23ec7))

### [0.1.1](https://github.com/gospime/passport-remember-me/compare/v0.1.0...v0.1.1) (2019-12-01)


### Bug Fixes

* set request into Issue callback, the same as for Verify ([5704274](https://github.com/gospime/passport-remember-me/commit/5704274d9c11b6a345e2e40b931052ff708cf916))

## [0.1.0](https://github.com/gospime/passport-remember-me/compare/v0.0.2...v0.1.0) (2019-12-01)


### ⚠ BREAKING CHANGES

* Added ability to set request into Verify callback. Added Passport-Strategy

### Features

* Added ability to set request into Verify callback. Added Passport-Strategy ([8b1ef80](https://github.com/gospime/passport-remember-me/commit/8b1ef80687d29860350239f9a0cfc984333d8291))

### [0.0.2](https://github.com/gospime/passport-remember-me/compare/v0.0.1...v0.0.2) (2019-12-01)


test/index.test.js
var rememberme = require('../src');


describe('passport-remember-me', function() {
    
  it('should export version', function() {
    expect(rememberme.version).to.be.a('string');
  });
    
  it('should export Strategy', function() {
    expect(rememberme.Strategy).to.be.a('function');
  });
  
});


src/strategy.js
/**
 * Module dependencies.
 */
var passport = require('passport-strategy')
  , util = require('util')
  , utils = require('./utils');


/**
 * `Strategy` constructor.
 *
 * @param {Object} options
 * @api public
 */
function Strategy(options, verify, issue) {
  if (typeof options == 'function') {
    issue = verify;
    verify = options;
    options = {};
  }
  if (!verify) throw new Error('remember me cookie authentication strategy requires a verify function');
  if (!issue) throw new Error('remember me cookie authentication strategy requires an issue function');
  
  var opts = { path: '/', httpOnly: true, maxAge: 604800000 }; // maxAge: 7 days
  this._key = options.key || 'remember_me';
  this._opts = utils.merge(opts, options.cookie);
  this._passReqToCallback = options.passReqToCallback;
  
  passport.Strategy.call(this);
  this.name = 'remember-me';
  this._verify = verify;
  this._issue = issue;
}

/**
 * Inherit from `passport.Strategy`.
 */
util.inherits(Strategy, passport.Strategy);

/**
 * Authenticate request based on remember me cookie.
 *
 * @param {Object} req
 * @api protected
 */
Strategy.prototype.authenticate = function(req, options) {
  // The rememeber me cookie is only consumed if the request is not
  // authenticated.  This is in preference to the session, which is typically
  // established at the same time the remember me cookie is issued.
  if (req.isAuthenticated()) { return this.pass(); }
  
  var token = null;
  if (req.cookies && typeof req.cookies === 'object') {
    token = req.cookies[this._key];
  } else {
    console.error('Maybe, you need to setup the package [cookie-parser](https://github.com/expressjs/cookie-parser)');
  }
  
  // Since the remember me cookie is primarily a convenience, the lack of one is
  // not a failure.  In this case, a response should be rendered indicating a
  // logged out state, rather than denying the request.
  if (!token) { return this.pass(); }
  
  var self = this;
  
  function verified(err, user, info) {
    if (err) { return self.error(err); }
    
    // Express exposes the response to the request.  We need the response to set
    // a cookie, so we'll grab it this way.  This breaks the encapsulation of
    // Passport's Strategy API, but is acceptable for this strategy.
    var res = req.res;
    
    if (!user) {
      // The remember me cookie was not valid.  However, because this
      // authentication method is primarily a convenience, we don't want to
      // deny the request.  Instead we'll clear the invalid cookie and proceed
      // to respond in a manner which indicates a logged out state.
      //
      // Note that a failure at this point may indicate a possible theft of the
      // cookie.  If handling this situation is a requirement, it is up to the
      // application to encode the value in such a way that this can be detected.
      // For a discussion on such matters, refer to:
      //   http://fishbowl.pastiche.org/2004/01/19/persistent_login_cookie_best_practice/
      //   http://jaspan.com/improved_persistent_login_cookie_best_practice
      //   http://web.archive.org/web/20130214051957/http://jaspan.com/improved_persistent_login_cookie_best_practice
      //   http://stackoverflow.com/questions/549/the-definitive-guide-to-forms-based-website-authentication
      
      res.clearCookie(self._key);
      return self.pass();
    }
    
    // The remember me cookie was valid and consumed.  For security reasons,
    // the just-used token should have been invalidated by the application.
    // A new token will be issued and set as the value of the remember me
    // cookie.
    function issued(err, val) {
      if (err) { return self.error(err); }
      res.cookie(self._key, val, self._opts);
      return self.success(user, info);
    }
    
    if (self._passReqToCallback) {
      self._issue(req, user, issued);
    } else {
      self._issue(user, issued);
    }
  }
  
  if (self._passReqToCallback) {
    self._verify(req, token, verified);
  } else {
    self._verify(token, verified);
  }
}


/**
 * Expose `Strategy`.
 */ 
module.exports = Strategy;
